eventdata/parameter_sources/interval_query_source.py

```diff
@@ -0,0 +1,136 @@
+    """Exception raised for parameter errors.
+
+    Attributes:
+        message -- explanation of the error
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+class ParameterSourceError(Error):
+    """Exception raised for runtime errors.
+
+    Attributes:
+        message -- explanation of the error
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+class IntervalQuerySource:
+    """
+    Adds range filter based on input from fieldstats_runner.
+
+    It expects the parameter hash to contain the following keys:
+
+
+        "body"                   -   Query body the range filter shouyld be added to. [Mandatory]
+        "index_pattern"          -   String representing the index pattern to query. This is used to look up correct field statistics. Defaults to 'filebeat-*'.
+        "type"                   -   Type to query within the index. Defaults to '*'.  
+        "fieldname"              -   Name of the field to filter on. Defaults to '@timestamp'.
+        "min_interval_size_pct"  -   Minimal percentage of the range to filter on. Parameter must be in the range 0 < min_interval_size_pct < 100. [Mandatory]
+        "max_interval_size_pct"  -   Maximum percentage of the range to filter on. Parameter must be in the range min_interval_size_pct < min_interval_size_pct <= 100. [Mandatory]
+        "cache"                  -   Boolean indicating whether request cache is to be used. Defaults to 'False'.
+
+    This parameter source will take the query supplied through the 'body' parameter and add a range filter to it. This will be based on the field specified
+    through the 'fieldname' parameter. The upper end of the range (range_max) will be set to the maximum value (fieldname_max) detected for the field through the fieldstats_runner.
+    The lower end of the range (range_min) will be set based on the following calculation:
+
+      range = fieldname_max - fieldname_min
+      interval_size_pct = min_interval_size_pct + random(max_interval_size_pct - min_interval_size_pct)
+      range_min = range_max - (range * interval_size_pct / 100)
+
+    """
+    def __init__(self, track, params, **kwargs):
+        self._index_pattern = "filebeat-*";
+        self._fieldname = "@timestamp";
+        self._type = "*";
+        self._cache = False;
+
+        random.seed();
+        if 'cache' in params.keys():
+            self._cache = params['cache']
+
+        if 'index_pattern' in params.keys():
+            self._index_pattern = params['index_pattern']
+
+        if 'type' in params.keys():
+            self._type = params['type']
+        
+        if 'fieldname' in params.keys():
+            self._fieldname = params['fieldname']
+
+        if 'body' in params.keys():
+            self._body = params['body']
+
+            # Verify that body contains `query.bool.must` key.
+            if not params['body']['query']['bool'] or ('must' not in params['body']['query']['bool'].keys()):
```

Comment:
This won't work because if a key in the structure is missing you're failing with a `KeyError`.

Reply:
Will change.